### PR TITLE
Global details on demand

### PR DESCRIPTION
**Pull request by ikesau**
https://user-images.githubusercontent.com/11844404/232628724-16c87c4e-e376-4f96-b244-88050e51feff.mov

### Overview:
1. Define all our DoDs in a [special Google Doc](https://docs.google.com/document/d/1pwTrbEKOy4D3LPbpbNaivQyYVv2KTePpw-P4F5fMNjA/edit)
2. Parse them into JSON and serve them all at `/dods.json`
3. Request the JSON in `runSiteFooterScripts` and attach it to the window
4. Add an event listener for any hover event
5. If the hovered element has the `dod-span` class and a `data-id` attribute that specifies a valid dod, show a tooltip.

### Questions

Are we okay with loading every single dod? 
**Pros**
- Far simpler to implement than tracking and attaching each dod per page.
- Could cache the JSON

**Cons**
- Currently we have 90 dods, averaging 400 bytes each. With ~300 dods we'd be up to 108KB, a ~3% increase in total pagesize.

### Remaining work to finish this:
- [x] Explorer support
- [x] Validation
- [x] Baking site
- [x] Baking static graphers
- [x] A migration of all the existing grapher charts to update their syntax to be the same as GDocs'
- [x] Export the current dods from the DB into the Gdoc
- [x] Style fixes
- [x] Strip out old dod code

Eventually we're also gonna want a cleaner way to handle these random properties on the `Gdoc` class. It's not sustainable to keep adding logic spaghetti to it for each new feature we want to solve with Gdocs.

### CODE REVIEW

**Comment by marcelgerber** [comment]
One thing I'm curious about is using a hashtag to denote DoD links inside Gdocs (e.g. `#dod-energy`).
I assume this is because Google does some URL validation, and will not allow invalid links being put in the field?
I'm asking, of course, because anchor links are already valid links, and we sometimes use them (mostly inside the yellow summary block, e.g. [here](https://ourworldindata.org/obesity)).

I suppose other syntaxes like `$dod-energy` or `%dod-energy` would not be allowed by Gdocs?

**Comment by marcelgerber** [comment]
Ah, and just as I wrote this I realize that the actual prefix is not `#`, but `#dod-`.
That's way better. In my mind, `#dod:` would be even better, if that works with GDocs.

File: db/model/Gdoc/archieToEnriched.ts [hunk 1]
```diff
@@ -177,5 +178,16 @@ export const archieToEnriched = (text: string): OwidGdocContent => {
             ? htmlToSimpleTextBlock(citation)
             : citation.map(htmlToSimpleTextBlock)
     parsed.toc = toc
+    if (parsed.details) {
+        parsed.details = keyBy(
+            parsed.details.map(
+                (detail: any): { id: string; text: EnrichedBlockText[] } => ({
+                    id: detail.id,
+                    text: detail.text.map(parseRawBlocksToEnrichedBlocks),
```
Comment (marcelgerber) [inline db/model/Gdoc/archieToEnriched.ts:186]: This means that we can, in theory, use full-fledged Archie syntax inside DoDs? Cool!

File: packages/@ourworldindata/utils/src/MarkdownTextWrap/MarkdownTextWrap.tsx [hunk 2]
```diff
@@ -296,9 +296,9 @@ export class IRDetailOnDemand extends IRElement {
     }
     toHTML(key?: React.Key): JSX.Element {
         return (
-            <DoDWrapper key={key} term={this.term} category={this.category}>
+            <a key={key} className="dod-span" data-id={this.term}>
```
Comment (marcelgerber) [inline packages/@ourworldindata/utils/src/MarkdownTextWrap/MarkdownTextWrap.tsx:299]: Looking at the gdoc, it seems we're getting rid now of the category/term distinction?

File: site/gdocs/utils.tsx [hunk 3]
```diff
@@ -71,6 +71,21 @@ export function renderSpan(
                 {renderSpans(span.children)}
             </a>
         ))
+        .with({ spanType: "span-dod" }, (span) => (
+            <span key={key}>
+                <a
+                    key={key}
```
Comment (marcelgerber) [inline site/gdocs/utils.tsx:77]: nit: don't need the inner key (or, alternatively, can get rid of the `<span>` wrapper)

File: site/globalDetailsOnDemand.scss [hunk 4]
```diff
@@ -0,0 +1,18 @@
+.global-dod {
```
Comment (marcelgerber) [inline site/globalDetailsOnDemand.scss:1]: I'm not sure this is a great name - in my thinking, we will use global DoD everywhere and have no other DoD implementation?
`dod-tooltip` would be a better name, then.

File: site/globalDetailsOnDemand.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        credentials: "same-origin",
+        headers: {
+            Accept: "application/json",
+        },
+    }).then((res) => res.json())
+
+    ;(window as any).details = details
+
+    document.addEventListener("mouseover", handleEvent)
+    document.addEventListener("touchstart", handleEvent)
```
Comment (marcelgerber) [inline site/globalDetailsOnDemand.tsx:20]: could add `{passive: true}` to these listeners to improve performance, see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

File: packages/@ourworldindata/utils/src/MarkdownTextWrap/MarkdownTextWrap.tsx (hunk 2 above)
Comment (ikesau) [reply packages/@ourworldindata/utils/src/MarkdownTextWrap/MarkdownTextWrap.tsx:299]: Yeah I don't think the category term thing really made an impact in the end. (I'll double check with the authors on that.)

If we have two different types of poverty that we want to define we can just treat them as separate terms: `poverty-growth` `poverty-health`

File: site/globalDetailsOnDemand.scss (hunk 4 above)
Comment (ikesau) [reply site/globalDetailsOnDemand.scss:1]: lol yeah the naming here was extremely "first thing that came to mind" 😛

once the pre-existing dod stuff is stripped out i'll be able to have better names

File: db/model/Gdoc/archieToEnriched.ts (hunk 1 above)
Comment (ikesau) [reply db/model/Gdoc/archieToEnriched.ts:186]: you can even have a recursive dod 🙈

File: settings/serverSettings.ts [hunk 6]
```diff
@@ -152,6 +152,9 @@ export const GDOCS_HOMEPAGE_CONFIG_DOCUMENT_ID: string =
 
 export const GDOCS_SHARED_DRIVE_ID = serverSettings.GDOCS_SHARED_DRIVE_ID ?? ""
 
+export const GDOCS_DETAILS_ON_DEMAND_ID =
```
Comment (ikesau) [inline settings/serverSettings.ts:155]: TODO: add error handling to help people if this variable is unset

Comment (marcelgerber) [reply settings/serverSettings.ts:155]: Ideally, our site would work "relatively" fine even if this setting is not set.

File: db/model/Gdoc/htmlToEnriched.ts [hunk 7]
```diff
@@ -120,7 +120,7 @@ export function cheerioToSpan(element: CheerioElement): Span | undefined {
                 if (className === "ref") {
                     return { spanType: "span-ref", children, url }
                 }
-                const dod = url.match(/#dod-(\w+)$/)
+                const dod = url.match(/#dod:(\w+)$/)
```
Comment (marcelgerber) [inline db/model/Gdoc/htmlToEnriched.ts:123]: `\w+` doesn't match any separators, like `-_/`.

File: site/globalDetailsOnDemand.tsx [hunk 8]
```diff
@@ -1,50 +1,63 @@
...
+export async function runGlobalDetailsOnDemand() {
+    const details: Record<string, Detail> = await fetch(
+        `${BAKED_BASE_URL}/dods.json`,
+        {
+            method: "GET",
+            credentials: "same-origin",
+            headers: {
+                Accept: "application/json",
+            },
+        }
```
Comment (marcelgerber) [inline site/globalDetailsOnDemand.tsx]: don't think you explicitly need to set these headers?

File: db/migration/1681860654784-UpdateDoDSyntax.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
+
+export class UpdateDoDSyntax1681860654784 implements MigrationInterface {
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`        
+            UPDATE charts SET config = JSON_SET(config, '$.subtitle', JSON_UNQUOTE(REGEXP_REPLACE(JSON_EXTRACT(config, '$.subtitle'), 'hover::[a-z]+::', '#dod:')));
```
[the remark this answers is not in the document]
Comment (ikesau) [reply db/migration/1681860654784-UpdateDoDSyntax.ts]: Ah okay 🙂 

Without exactly understanding the usecase for these rows, I added the migration for: 

```
charts.config
chart_revisions.config
suggested_chart_revisions.suggestedConfig
suggested_chart_revisions.originalConfig
```

Lmk if that's correct!

File: settings/serverSettings.ts (hunk 6 above)
Comment (ikesau) [reply settings/serverSettings.ts:155]: Hopefully this time I'll have learned my lesson from all the people needing help in slack with the last round of Gdocs variables 😅

File: baker/SiteBaker.tsx [hunk 10]
```diff
@@ -375,6 +378,46 @@ export class SiteBaker {
...
+            return
+        }
+
+        const charts: { slug: string; subtitle: string; note: string }[] =
+            await db.queryMysql(`
+            SELECT config->>'$.slug' as slug, config->>'$.subtitle' as subtitle, config->>'$.note' as note
+            FROM charts
+            WHERE JSON_EXTRACT(config, "$.isPublished") = true
+            AND JSON_EXTRACT(config, "$.subtitle") LIKE "%#dod:%"
+            OR JSON_EXTRACT(config, "$.note") LIKE "%#dod:%"
```
Comment (marcelgerber) [inline baker/SiteBaker.tsx:407]: Probably best to wrap these two lines into parens.

File: packages/@ourworldindata/utils/src/MarkdownTextWrap/parser.ts [hunk 11]
```diff
@@ -326,7 +325,11 @@ const detailOnDemandContentParser: (
         r.nonBracketWord
     )
 
-export const detailOnDemandRegex = /\(hover::(\w+)::(\w+)\)/
+export const detailOnDemandRegex = /\(#dod:(\w+)\)/g
```
Comment (marcelgerber) [inline packages/@ourworldindata/utils/src/MarkdownTextWrap/parser.ts:328]: This is a different regex than in `htmlToEnriched` 😱

Comment (ikesau) [reply packages/@ourworldindata/utils/src/MarkdownTextWrap/parser.ts]: Oops! Extracted them to a single value in `utils/src/GdocsUtils.ts`

File: packages/@ourworldindata/utils/src/GdocsUtils.ts [hunk 12]
```diff
@@ -7,6 +7,10 @@ import { OwidGdocLinkJSON } from "./owidTypes.js"
 // https://docs.google.com/document/u/0/d/abcd-1234/edit?usp=sharing
 export const gdocUrlRegex = /https:\/\/docs\.google\.com\/.+?\/([-\w]+)\/edit/
 
+// Works for:
```
Comment (marcelgerber) [inline packages/@ourworldindata/utils/src/GdocsUtils.ts:14]: nit: unfinished comment

File: baker/SiteBaker.tsx (hunk 10 above)
Comment (marcelgerber) [reply baker/SiteBaker.tsx]: ah, no, what I actually meant here is to wrap the `JSON_EXTRACT() OR JSON_EXTRACT()` into parens.

File: db/model/Gdoc/Gdoc.ts [hunk 13]
```diff
@@ -413,6 +457,17 @@ export class Gdoc extends BaseEntity implements OwidGdocInterface {
         return gdoc
     }
 
+    static async getDetailsOnDemandGdoc(): Promise<
+        DetailDictionary | undefined
+    > {
+        const gdoc = await Gdoc.getGdocFromContentSource(
```
Comment (ikesau) [inline db/model/Gdoc/Gdoc.ts:463]: This should have an error handler if `GDOCS_DETAILS_ON_DEMAND_ID` is undefined

Comment (ikesau) [reply db/model/Gdoc/Gdoc.ts:463]: Also double check what causes the error when adding this document the first time

Comment (ikesau) [reply db/model/Gdoc/Gdoc.ts:463]: And remove the old details button from the sidebar

**Comment by ikesau** [comment]
Todo:
let daniel know as i merge so the datasette import can be fixed
better validation on the dod document
include better documentation in the document (explain how publishing works and allowed content types)

File: public/_headers [hunk 14]
```diff
@@ -14,6 +14,8 @@
 
 /headerMenu.json
     Access-Control-Allow-Origin: *
+    Access-Control-Allow-Headers: content-type
```
Comment (marcelgerber) [inline public/_headers:17]: Do we actually need these? Like, does it not work if these headers are not present?

**Comment by ikesau** [comment]
@danyx23 okay I've added some amount of validation and logging 🙂 

We don't currently have a nice flow for validating non-article documents, so it's still possible to publish the Dod document with invalid data (which we would then catch and log during baking instead)

I could add something bespoke for the Dod document to accomplish this, but it would be kind of janky and one-off.

If it's okay with you, I'd prefer to wait until we can implement a validation abstraction for all types of fragments.

**Comment by ikesau** [comment]
https://github.com/owid/owid-grapher-svgs/pull/11 should be merged when this PR is merged

**Pull request merged**